src/client/client.rs

```diff
@@ -445,91 +419,102 @@ where
         let is_ver_h2 = ver == Ver::Http2;
         let connector = self.connector.clone();
         let dst = domain_as_uri(pool_key.clone());
-        hyper_lazy(move || {
-            // Try to take a "connecting lock".
-            //
-            // If the pool_key is for HTTP/2, and there is already a
-            // connection being established, then this can't take a
-            // second lock. The "connect_to" future is Canceled.
-            let connecting = match pool.connecting(&pool_key, ver) {
-                Some(lock) => lock,
-                None => {
-                    let canceled =
-                        crate::Error::new_canceled().with("HTTP/2 connection in progress");
-                    return Either::Right(future::err(canceled));
-                }
-            };
-            Either::Left(
-                connector
-                    .connect(connect::sealed::Internal, dst)
-                    .map_err(crate::Error::new_connect)
-                    .and_then(move |io| {
-                        let connected = io.connected();
-                        // If ALPN is h2 and we aren't http2_only already,
-                        // then we need to convert our pool checkout into
-                        // a single HTTP2 one.
-                        let connecting = if connected.alpn == Alpn::H2 && !is_ver_h2 {
-                            match connecting.alpn_h2(&pool) {
-                                Some(lock) => {
-                                    trace!("ALPN negotiated h2, updating pool");
-                                    lock
-                                }
-                                None => {
-                                    // Another connection has already upgraded,
-                                    // the pool checkout should finish up for us.
-                                    let canceled = crate::Error::new_canceled()
-                                        .with("ALPN upgraded to HTTP/2");
-                                    return Either::Right(future::err(canceled));
-                                }
+        // Try to take a "connecting lock".
+        //
+        // If the pool_key is for HTTP/2, and there is already a
+        // connection being established, then this can't take a
+        // second lock. The "connect_to" future is Canceled.
+        let connecting = match pool.connecting(&pool_key, ver) {
+            Some(lock) => lock,
+            None => {
+                let canceled = crate::Error::new_canceled().with("HTTP/2 connection in progress");
+                return Either::Right(future::err(canceled));
+            }
+        };
+        Either::Left(Box::pin(async move {
```

Comment:
Ok, I see, by putting the call to `connector.connect()` inside this sub future, `async move`, it won't call `connect()` until the first poll. Got it.